src/main/scala/FundingMode/Calculation.scala

```diff
@@ -0,0 +1,30 @@
+package FundingMode
+
+import FundingMode.CaseClassDeclaration._
+
+object Calculation {
+  val prohibited  = "prohibited"
+  val collections = "collections"
+  val receipts    = "receipts"
+
+  val from_client          = "from_client"
+  val obo_client           = "obo_client"
+  val obo_clients_customer = "obo_clients_customer"
+
+  def identifyFundingType(resolver: AccountResolver, sender: Sender): String = {
+    if (resolver.isNotComplianceRelationship || resolver.isNestedPaymentsWithCollections || resolver.isRegulatedAffiliateReceipts) prohibited
+    else if (resolver.isCorporateCollections || sender.isNotAccountHolder || sender.isApprovedFundingPartner) collections
+    else if (sender.isAccountHolder) receipts
+    else throw new IllegalArgumentException("Can't resolve funding_type")
```

Comment:
Avoid exception at all. Try to define strict method's protocol and avoid if else structures. in the end this method can look like:

def identifyFundingType(resolver: AccountResolver, sender: Sender): Option[FundingyType])= (resolver, sender) match{
case (res, _) if (res.isNotComplianceRelationship || res.isNestedPaymentsWithCollections || res.isRegulatedAffiliateReceipts) => Some(Prohibited)
case(res, s) if (res.isCorporateCollections || s.isNotAccountHolder || s.isApprovedFundingPartner) => Some(Collections)
case (_,s) if s. sender.isAccountHolder => Some(Receipts)
case _=> None
}
Now you can clearly see what is going on, if you able to identify type you get Some(type) if not you get None